pilopt/src/qse_opt.rs

```diff
@@ -0,0 +1,135 @@
+            Identity::Polynomial(PolynomialIdentity { expression, .. }) => {
+                try_algebraic_to_quadratic_symbolic_expression(expression)
+            }
+            _ => None,
+        })
+        .collect_vec();
+
+    match solver::Solver::new(identities).solve() {
+        Err(_) => {
+            log::error!("Error while QSE-optimizing. This is usually the case when the constraints are inconsistent.");
+        }
+        Ok(SolveResult {
+            simplified_algebraic_constraints,
+            ..
+        }) => {
+            let mut replacements = 0;
+            // TODO this is not correct if the are constraints with publics and challenges, because we skip them.
+            for identity in &mut pil_file.identities {
+                if let Identity::Polynomial(PolynomialIdentity { expression, .. }) = identity {
+                    if let Some(constraint) = simplified_algebraic_constraints.get(replacements) {
+                        *expression = quadratic_symbolic_expression_to_algebraic(constraint);
+                    } else {
+                        *expression = AlgebraicExpression::Number(0.into());
+                    }
+                    replacements += 1;
+                }
+            }
+            // If not, we have more simplified algebraic constraints than original
+            // and would have to append.
+            assert!(replacements >= simplified_algebraic_constraints.len());
+        }
+    }
+}
+
+// Turns an algebraic expression into a quadratic symbolic expression,
+/// assuming all [`AlgebraicReference`]s are unknown variables.
+///
+/// Fails (returns None) if public references or challenges are used.
+fn try_algebraic_to_quadratic_symbolic_expression<T: FieldElement>(
+    expr: &AlgebraicExpression<T>,
+) -> Option<QuadraticSymbolicExpression<T, AlgebraicReference>> {
+    Some(match expr {
+        AlgebraicExpression::Reference(r) => {
+            QuadraticSymbolicExpression::from_unknown_variable(r.clone())
+        }
+        AlgebraicExpression::Number(n) => (*n).into(),
+        AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation { left, op, right }) => {
+            let left = try_algebraic_to_quadratic_symbolic_expression(left)?;
+            let right = try_algebraic_to_quadratic_symbolic_expression(right)?;
+            match op {
+                AlgebraicBinaryOperator::Add => left + right,
+                AlgebraicBinaryOperator::Sub => left - right,
+                AlgebraicBinaryOperator::Mul => left * right,
+                AlgebraicBinaryOperator::Pow => {
+                    todo!()
+                }
+            }
+        }
+        AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op, expr }) => match op {
+            AlgebraicUnaryOperator::Minus => -try_algebraic_to_quadratic_symbolic_expression(expr)?,
+        },
+        AlgebraicExpression::PublicReference(_) | AlgebraicExpression::Challenge(_) => return None,
```

Comment:
This should be an easy one to fix though, right? Just use an enum (algebraic ref, public ref, challenge) as the variable type?